autoPlay.user.js

```diff
@@ -956,6 +970,125 @@ function disableCooldownIfRelevant() {
+	var oddsOfElement = 1 - (0.75 * 0.75 * 0.75);
+
+	var buyUpgrade = function(id) {
+		// If upgrade is element damage
+		if (id >= 3 && 6 >= id) {
+			s().TryUpgrade(document.getElementById('upgr_' + id).childElements()[3]);
+		} else {
+			s().TryUpgrade(document.getElementById('upgr_' + id).childElements()[0].childElements()[1]);
+		}
+	};
+
+	var myGold = s().m_rgPlayerData.gold;
+
+	// Initial values for armor and click damage 
+	var bestUpgradeForDamage, bestUpgradeForArmor;
+	var highestUpgradeValueForDamage = 0;
+	var highestUpgradeValueForArmor = 0;
+	var highestElementLevel = 0;
+	var purchasedShieldsWhileRespawning = false;
+
+	var critMultiplier = s().m_rgPlayerTechTree.damage_multiplier_crit;
+	var critChance = s().m_rgPlayerTechTree.crit_percentage;
+	var dpc = s().m_rgPlayerTechTree.damage_per_click;
+
+	var upgradeCost;
+
+	var upgrades = s().m_rgTuningData.upgrades.slice(0);
+
+	for (var i=0; i< upgrades.length; i++) {
+		var upgrade = upgrades[i];
+
+		if (upgrade.required_upgrade !== undefined)
+		{
+			var requiredUpgradeLevel = upgrade.required_upgrade_level !== undefined ? upgrade.required_upgrade_level : 1;
+			var parentUpgradeLevel = s().GetUpgradeLevel(upgrade.required_upgrade);
+			if (requiredUpgradeLevel > parentUpgradeLevel) {
+				// If upgrade is not available, we skip it
+				continue;
+			}
+		}
+
+		var upgradeCurrentLevel = s().m_rgPlayerUpgrades[i].level;
+		upgradeCost = s().m_rgPlayerUpgrades[i].cost_for_next_level;
+
+		switch(upgrade.type) {
+			case UPGRADE_TYPES.ARMOR:
+				// HP increase per cost
+				if (upgrade.multiplier / upgradeCost > highestUpgradeValueForArmor) {
+					bestUpgradeForArmor = i;
+					highestUpgradeValueForArmor = upgrade.multiplier / upgradeCost;
+				}
+				break;
+			case UPGRADE_TYPES.CLICK_DAMAGE:
+				// Damage increase per cost
+				if ((critChance * critMultiplier + 1) * currentClickRate * upgrade.multiplier / upgradeCost > highestUpgradeValueForDamage) {
+					bestUpgradeForDamage = i;
+					highestUpgradeValueForDamage = (critChance * critMultiplier + 1) * currentClickRate * upgrade.multiplier / upgradeCost;
+				}
+				break;
+			case UPGRADE_TYPES.DPS:
+				// Damage increase per cost
+				if (upgrade.multiplier / upgradeCost > highestUpgradeValueForDamage) {
+					bestUpgradeForDamage = i;
+					highestUpgradeValueForDamage = upgrade.multiplier / upgradeCost;
+				}
+				break;
+			case UPGRADE_TYPES.ELEMENTAL_FIRE:
+			case UPGRADE_TYPES.ELEMENTAL_WATER:
+			case UPGRADE_TYPES.ELEMENTAL_AIR:
+			case UPGRADE_TYPES.ELEMENTAL_EARTH:
+				break;
+			case UPGRADE_TYPES.LUCKY_SHOT:
+				if (upgrade.multiplier * dpc * critChance * currentClickRate / upgradeCost > highestUpgradeValueForDamage) { // dmg increase per moneys
+					bestUpgradeForDamage = i;
+					highestUpgradeValueForDamage = upgrade.multiplier * dpc * critChance * currentClickRate / upgradeCost;
+				}
+				break;
+			default:
+				break;
+		}
+	}
+
+	var currentHealth = s().CurrentScene().m_rgPlayerData.hp;
```

Comment:
This would cause runtime issues, since CurrentScene() is not a function, proper implementation would be ```s().m_rgPlayerData.hp```